lib/icontrol_rest/client.rb

```diff
@@ -64,38 +64,51 @@ def logger
     #         "nameServers"=>["1.2.3.72", "1.2.3.73"],
     #         "search"=>["domain.com"]}
     #
-    %w[delete get].each do |action|
-      define_method(action.to_s) do |route, _arg = nil|
-        send_request { self.class.send(action, route, @options) }
+    %w[delete get post put patch].each do |action|
+      define_method(action.to_s) do |route:, headers: {}|
+        send_request(action: action, route: route, headers: headers)
       end
     end
 
-    # Public  - Post/Put request methods.
+    # Public    - Post/Put/Patch request methods.
     #
-    # route   - URI to send to server as a string.
-    # data    - Hash of parameters to pass with request.
+    # route     - URI to send to server as a string.
+    # body      - Hash representing the request body.
+    # headers   - Hash of headers to pass with request.
     #
-    # Returns - response from the server as a hash.
+    # Returns   - response from the server as a hash.
     %w[post put patch].each do |action|
-      define_method(action.to_s) do |route, data|
-        body = {}
-        body[:body] = data.to_json
-        send_request { self.class.send(action, route, body.merge(@options)) }
+      define_method(action.to_s) do |route:, body: {}, headers: {}|
+        send_request(action: action, route: route, body: body, headers: headers)
       end
     end
 
     private
 
+    # Private - Builds a Hash of headers for a request.
+    #
+    # headers - A Hash of headers to be included with the required headers for an iControl request.
+    #
+    # Returns - A Hash of required headers and any additional headers to be included in the request.
+    def build_headers(headers:)
```

Comment:
What's the benefit of moving the header structure into this method?

Reply:
Removed this. Was able to refactor a bit to keep this the way it was and still support a user to pass headers.